base/strings/string.jl

```diff
@@ -139,6 +139,111 @@ end
 typemin(::Type{String}) = ""
 typemin(::String) = typemin(String)
 
+##
+#=
+  ┌─────────────────────────────────────────────────────┐
+  │                 Forward Mode State Diagram          │
+  │  INCLUSIVE    ┌──────────────2──────────────┐       │
+  │    UTF-8      │                             │       │
+  │               ├────────3────────┐           │       │
+  │   IUTF-8      │                 │           │       │
+  │    ┌─0─┐      │     ┌─┐        ┌▼┐         ┌▼┐      │
+  │    │   │      ├─4──►│3├───1────►2├────1────►1├────┐ │
+  │   ┌▼───┴┐     │     └─┘        └─┘         └─┘    │ │
+  │   │  0  ├─────┘   Needs 3    Needs 2     Needs 1  │ │
+  │   └───▲─┘        ContBytes  ContBytes   ContBytes │ │
+  │       │                                           │ │
+  │       │           ContByte=Transition 1           │ │
+  │       └─────────────────────1─────────────────────┘ │
+  │  ┌─┐                                                │
+  │  │4│◄───All undefined transitions result in state 4 │
+  │  └─┘      State machine must be reset after state 4 │
+  └─────────────────────────────────────────────────────┘
+
+  ┌─────────────────────────────────────────────────────┐
+  │                 Reverse Mode State Diagram          │
+  │  INCLUSIVE    ┌──◄───────────2:4────────────┐       │
+  │    UTF-8      │                             │       │
+  │   IUTF-8      ├──◄─────3:4──────┐           │       │
+  │               │                 │           │       │
+  │  ┌─0,2:4─┐    │     ┌─┐        ┌┴┐         ┌┴┐      │
+  │  │       │    ├─4───┤3│◄──1────┤2│◄───1────┤1│◄───┐ │
+  │ ┌▼───────┴┐   │     └─┘        └─┘         └─┘    │ │
+  │ │    0    │◄──┘   Needs 3    Needs 2     Needs 1  │ │
+  │ └─────┬───┘      ContBytes  ContBytes   ContBytes │ │
+  │       │                                           │ │
+  │       │           ContByte=Transition 1           │ │
+  │       └─────────────────────1─────────────────────┘ │
+  │  ┌─┐                                                │
+  │  │4│◄───All undefined transitions result in state 4 │
+  │  └─┘      State machine must be reset after state 4 │
+  └─────────────────────────────────────────────────────┘
+=#
+const _IUTF8State = UInt16
+const _IUTF8_SHIFT_MASK = _IUTF8State(0b1111)
+const _IUTF8_DFA_ACCEPT = _IUTF8State(0)
+const _IUTF8_DFA_INVALID = _IUTF8State(4)
+
+const _IUTF8_DFA_TABLE, _IUTF8_DFA_REVERSE_TABLE = begin
```

Comment:
You probably want to use `let` here, otherwise all temporary variables pollute the module
```suggestion
const _IUTF8_DFA_TABLE, _IUTF8_DFA_REVERSE_TABLE = let
```